### DISCUSSION TITLE
Renovate bot fails when there are nested go projects without a go.mod

### DISCUSSION THREAD

**Discussion by jorgemmsilva** [renovatebot/renovate#20066] [Request Help] [answered]
### I'm using the Mend Renovate hosted app.

- [X] I'm using the Mend Renovate hosted app.

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe your problem.

https://github.com/iotaledger/wasp/pull/1737

As it can be seen in this comment: https://github.com/iotaledger/wasp/pull/1737#issuecomment-1402579480

We have some directories that have go code, but no go.mod (those are destined to be compiled using tinygo).

`go get -d -t ./...` seems to be broken when the bot tries to run it. The same command works fine in a local clone of the repo.

### Relevant debug logs.

<details><summary>Logs</summary>
Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v "/mnt/renovate/gh/iotaledger/wasp":"/mnt/renovate/gh/iotaledger/wasp" -v "/tmp/renovate-cache":"/tmp/renovate-cache" -v "/tmp/containerbase":"/tmp/containerbase" -e GOPATH -e GOPROXY -e GOFLAGS -e CGO_ENABLED -e GIT_CONFIG_KEY_0 -e GIT_CONFIG_VALUE_0 -e GIT_CONFIG_KEY_1 -e GIT_CONFIG_VALUE_1 -e GIT_CONFIG_KEY_2 -e GIT_CONFIG_VALUE_2 -e GIT_CONFIG_COUNT -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w "/mnt/renovate/gh/iotaledger/wasp" docker.io/renovate/sidecar bash -l -c "install-tool golang 1.19.5 && go get -d -t ./..."
go: downloading github.com/iotaledger/go-ethereum v1.10.26-wasp
go: downloading github.com/iotaledger/iota.go/v3 v3.0.0-rc.1.0.20221212191201-0105a0847e96
go: downloading github.com/iotaledger/iota.go v1.0.0
go: downloading github.com/stretchr/testify v1.8.1
go: downloading github.com/iotaledger/hive.go/serializer/v2 v2.0.0-rc.1
go: downloading github.com/iotaledger/hive.go/core v1.0.0-rc.2.0.20230120131747-b65c9cb22ea9
go: downloading github.com/iotaledger/inx-app v1.0.0-rc.2
go: downloading go.uber.org/dig v1.16.1
go: downloading github.com/golang-jwt/jwt v3.2.2+incompatible
go: downloading github.com/labstack/echo/v4 v4.10.0
go: downloading golang.org/x/exp v0.0.0-20230126173853-a67bb567ff2e
go: downloading pgregory.net/rapid v0.5.5
go: downloading github.com/kape1395/kyber/v3 v3.0.14-0.20230124095845-ec682ff08c93
go: downloading go.uber.org/zap v1.24.0
go: downloading go.uber.org/atomic v1.10.0
go: downloading github.com/samber/lo v1.37.0
go: downloading github.com/prometheus/client_golang v1.14.0
go: downloading golang.org/x/crypto v0.5.0
go: downloading github.com/PuerkitoBio/goquery v1.8.0
go: downloading github.com/hashicorp/golang-lru v0.6.0
go: downloading github.com/iotaledger/inx/go v1.0.0-rc.1
go: downloading github.com/libp2p/go-libp2p v0.24.2
go: downloading github.com/multiformats/go-multiaddr v0.8.0
go: downloading nhooyr.io/websocket v1.8.7
go: downloading go.nanomsg.org/mangos/v3 v3.4.2
go: downloading github.com/bytecodealliance/wasmtime-go/v5 v5.0.0
go: downloading github.com/second-state/WasmEdge-go v0.11.2
go: downloading github.com/wasmerio/wasmer-go v1.0.4
go: downloading github.com/pangpanglabs/echoswagger/v2 v2.4.1
go: downloading github.com/bygui86/multi-profile/v2 v2.1.0
go: downloading github.com/labstack/echo-contrib v0.13.1
go: downloading go.elastic.co/apm/module/apmechov4 v1.15.0
go: downloading go.elastic.co/apm v1.15.0
go: downloading github.com/spf13/pflag v1.0.5
go: downloading gopkg.in/yaml.v3 v3.0.1
go: downloading google.golang.org/grpc v1.52.0
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/pmezard/go-difflib v1.0.0
go: downloading github.com/hashicorp/go-version v1.6.0
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/go-stack/stack v1.8.1
go: downloading github.com/grpc-ecosystem/go-grpc-middleware v1.3.0
go: downloading github.com/grpc-ecosystem/go-grpc-prometheus v1.2.0
go: downloading github.com/knadh/koanf v1.5.0
go: downloading github.com/spf13/cast v1.5.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/labstack/gommon v0.4.0
go: downloading golang.org/x/net v0.5.0
go: downloading github.com/valyala/fasttemplate v1.2.2
go: downloading golang.org/x/time v0.3.0
go: downloading github.com/cockroachdb/errors v1.9.0
go: downloading github.com/mr-tron/base58 v1.2.0
go: downloading golang.org/x/xerrors v0.0.0-20220907171357-04be3eba64a2
go: downloading go.uber.org/multierr v1.9.0
go: downloading github.com/stretchr/objx v0.5.0
go: downloading github.com/beorn7/perks v1.0.1
go: downloading github.com/cespare/xxhash/v2 v2.2.0
go: downloading github.com/golang/protobuf v1.5.2
go: downloading github.com/prometheus/client_model v0.3.0
go: downloading github.com/prometheus/common v0.39.0
go: downloading github.com/prometheus/procfs v0.9.0
go: downloading golang.org/x/sys v0.4.0
go: downloading google.golang.org/protobuf v1.28.1
go: downloading github.com/oasisprotocol/ed25519 v0.0.0-20210505154701-76d8c688d86e
go: downloading go.dedis.ch/fixbuf v1.0.3
go: downloading github.com/andybalholm/cascadia v1.3.1
go: downloading github.com/pelletier/go-toml/v2 v2.0.6
go: downloading github.com/pelletier/go-toml v1.9.5
go: downloading github.com/iotaledger/grocksdb v1.7.5-0.20221128103803-fcdb79760195
go: downloading go.dedis.ch/protobuf v1.0.11
go: downloading github.com/btcsuite/btcd/btcec/v2 v2.3.2
go: downloading github.com/deckarep/golang-set v1.8.0
go: downloading github.com/gorilla/websocket v1.5.0
go: downloading gopkg.in/natefinch/npipe.v2 v2.0.0-20160621034901-c1b8fa8bdcce
go: downloading github.com/eclipse/paho.mqtt.golang v1.4.2
go: downloading github.com/pasztorpisti/qs v0.0.0-20171216220353-8d6c33ee906c
go: downloading github.com/panjf2000/ants/v2 v2.7.1
go: downloading github.com/multiformats/go-multiaddr-dns v0.3.1
go: downloading go.uber.org/fx v1.19.1
go: downloading github.com/decred/dcrd/dcrec/secp256k1/v4 v4.1.0
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading github.com/libp2p/go-openssl v0.1.0
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/ipfs/go-cid v0.3.2
go: downloading github.com/multiformats/go-multicodec v0.7.0
go: downloading github.com/multiformats/go-multihash v0.2.1
go: downloading github.com/ipfs/go-log/v2 v2.5.1
go: downloading github.com/lucas-clemente/quic-go v0.31.1
go: downloading github.com/multiformats/go-multiaddr-fmt v0.1.0
go: downloading github.com/libp2p/go-reuseport v0.2.0
go: downloading github.com/marten-seemann/tcp v0.0.0-20210406111302-dfbc87cc63fd
go: downloading github.com/mikioh/tcpinfo v0.0.0-20190314235526-30a79bb1804b
go: downloading github.com/multiformats/go-multibase v0.1.1
go: downloading github.com/multiformats/go-varint v0.0.7
go: downloading github.com/klauspost/compress v1.15.14
go: downloading github.com/edsrzf/mmap-go v1.1.0
go: downloading github.com/holiman/uint256 v1.2.1
go: downloading github.com/dustin/go-humanize v1.0.1
go: downloading go.elastic.co/apm/module/apmhttp v1.15.0
go: downloading filippo.io/edwards25519 v1.0.0
go: downloading github.com/tcnksm/go-latest v0.0.0-20170313132115-e3007ae9052e
go: downloading github.com/sasha-s/go-deadlock v0.3.1
go: downloading google.golang.org/genproto v0.0.0-20230117162540-28d6b9783ac4
go: downloading github.com/mitchellh/copystructure v1.2.0
go: downloading github.com/mitchellh/mapstructure v1.5.0
go: downloading github.com/fsnotify/fsnotify v1.6.0
go: downloading github.com/fatih/structs v1.1.0
go: downloading github.com/mattn/go-colorable v0.1.13
go: downloading github.com/mattn/go-isatty v0.0.17
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/cockroachdb/logtags v0.0.0-20230118201751-21c54148d20b
go: downloading github.com/cockroachdb/redact v1.1.3
go: downloading github.com/getsentry/sentry-go v0.17.0
go: downloading github.com/iancoleman/orderedmap v0.2.0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.4
go: downloading github.com/golang/snappy v0.0.4
go: downloading github.com/olekukonko/tablewriter v0.0.5
go: downloading github.com/prometheus/tsdb v0.10.0
go: downloading github.com/VictoriaMetrics/fastcache v1.12.0
go: downloading github.com/holiman/bloomfilter/v2 v2.0.3
go: downloading github.com/shirou/gopsutil v3.21.11+incompatible
go: downloading github.com/syndtr/goleveldb v1.0.1-0.20210819022825-2ae1ddf74ef7
go: downloading golang.org/x/sync v0.1.0
go: downloading github.com/emirpasic/gods v1.18.1
go: downloading github.com/libp2p/go-flow-metrics v0.1.0
go: downloading github.com/libp2p/go-msgio v0.3.0
go: downloading github.com/benbjohnson/clock v1.3.0
go: downloading github.com/libp2p/go-netroute v0.2.1
go: downloading github.com/multiformats/go-multistream v0.3.3
go: downloading github.com/pbnjay/memory v0.0.0-20210728143218-7b4eea64cf58
go: downloading github.com/libp2p/go-yamux/v4 v4.0.0
go: downloading github.com/raulk/go-watchdog v1.3.0
go: downloading github.com/jbenet/go-temp-err-catcher v0.1.0
go: downloading github.com/libp2p/go-buffer-pool v0.1.0
go: downloading github.com/libp2p/go-libp2p-asn-util v0.2.0
go: downloading github.com/flynn/noise v1.0.0
go: downloading github.com/google/gopacket v1.1.19
go: downloading github.com/miekg/dns v1.1.50
go: downloading github.com/mattn/go-pointer v0.0.1
go: downloading github.com/spacemonkeygo/spacelog v0.0.0-20180420211403-2296661a0572
go: downloading github.com/klauspost/cpuid/v2 v2.2.3
go: downloading github.com/golang/mock v1.6.0
go: downloading github.com/onsi/ginkgo/v2 v2.7.0
go: downloading github.com/mikioh/tcpopt v0.0.0-20190314235656-172688c1accc
go: downloading github.com/multiformats/go-base32 v0.1.0
go: downloading github.com/multiformats/go-base36 v0.2.0
go: downloading github.com/Microsoft/go-winio v0.6.0
go: downloading github.com/elastic/go-licenser v0.3.1
go: downloading github.com/elastic/go-sysinfo v1.1.1
go: downloading github.com/jcchavezs/porto v0.1.0
go: downloading github.com/santhosh-tekuri/jsonschema v1.2.4
go: downloading go.elastic.co/fastjson v1.1.0
go: downloading golang.org/x/lint v0.0.0-20210508222113-6edffad5e616
go: downloading golang.org/x/tools v0.5.0
go: downloading github.com/google/go-github v17.0.0+incompatible
go: downloading github.com/kr/text v0.2.0
go: downloading github.com/petermattis/goid v0.0.0-20221215004737-a150e88a970d
go: downloading github.com/mitchellh/reflectwalk v1.0.2
go: downloading golang.org/x/text v0.6.0
go: downloading github.com/kr/pretty v0.3.1
go: downloading github.com/mattn/go-runewidth v0.0.14
go: downloading github.com/tklauser/go-sysconf v0.3.11
go: downloading github.com/yusufpapurcu/wmi v1.2.2
go: downloading github.com/huin/goupnp v1.0.3
go: downloading github.com/jackpal/go-nat-pmp v1.0.2
go: downloading github.com/libp2p/go-nat v0.1.0
go: downloading github.com/containerd/cgroups v1.0.4
go: downloading github.com/elastic/gosigar v0.14.2
go: downloading github.com/davidlazar/go-crypto v0.0.0-20200604182044-b73af7476f6c
go: downloading github.com/libp2p/go-cidranger v1.1.0
go: downloading github.com/francoispqt/gojay v1.2.13
go: downloading golang.org/x/mod v0.7.0
go: downloading github.com/armon/go-radix v1.0.0
go: downloading github.com/google/go-querystring v1.1.0
go: downloading github.com/rogpeppe/go-internal v1.9.0
go: downloading github.com/rivo/uniseg v0.4.3
go: downloading github.com/tklauser/numcpus v0.6.0
go: downloading github.com/go-ole/go-ole v1.2.6
go: downloading github.com/koron/go-ssdp v0.0.3
go: downloading github.com/coreos/go-systemd/v22 v22.5.0
go: downloading github.com/docker/go-units v0.5.0
go: downloading github.com/godbus/dbus/v5 v5.1.0
go: downloading github.com/opencontainers/runtime-spec v1.0.2
go: downloading lukechampine.com/blake3 v1.1.7
go: downloading github.com/spaolacci/murmur3 v1.1.0
go: downloading github.com/marten-seemann/qtls-go1-18 v0.1.4
go: downloading github.com/marten-seemann/qtls-go1-19 v0.1.2
go: downloading github.com/go-task/slim-sprig v0.0.0-20210107165309-348f09dbbbc0
go: downloading github.com/joeshaw/multierror v0.0.0-20140124173710-69b34d4ec901
go: downloading howett.net/plist v0.0.0-20181124034731-591f970eefbb
go: downloading github.com/elastic/go-windows v1.0.0
go: downloading github.com/google/pprof v0.0.0-20230111200839-76d1ae5aea2b
github.com/iotaledger/wasp/contracts/wasm/dividend/go/dividendimpl imports
	github.com/iotaledger/wasp/contracts/wasm/dividend/go/dividend: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/dividend/go/dividend
github.com/iotaledger/wasp/contracts/wasm/donatewithfeedback/go/donatewithfeedbackimpl imports
	github.com/iotaledger/wasp/contracts/wasm/donatewithfeedback/go/donatewithfeedback: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/donatewithfeedback/go/donatewithfeedback
github.com/iotaledger/wasp/contracts/wasm/erc721/go/erc721impl imports
	github.com/iotaledger/wasp/contracts/wasm/erc721/go/erc721: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/erc721/go/erc721
github.com/iotaledger/wasp/contracts/wasm/fairauction/go/fairauctionimpl imports
	github.com/iotaledger/wasp/contracts/wasm/fairauction/go/fairauction: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/fairauction/go/fairauction
github.com/iotaledger/wasp/contracts/wasm/fairroulette/go/fairrouletteimpl imports
	github.com/iotaledger/wasp/contracts/wasm/fairroulette/go/fairroulette: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/fairroulette/go/fairroulette
github.com/iotaledger/wasp/contracts/wasm/inccounter/go/inccounterimpl imports
	github.com/iotaledger/wasp/contracts/wasm/inccounter/go/inccounter: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/inccounter/go/inccounter
github.com/iotaledger/wasp/contracts/wasm/testcore/go/testcoreimpl imports
	github.com/iotaledger/wasp/contracts/wasm/testcore/go/testcore: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/testcore/go/testcore
github.com/iotaledger/wasp/contracts/wasm/testwasmlib/go/testwasmlibimpl imports
	github.com/iotaledger/wasp/contracts/wasm/testwasmlib/go/testwasmlib: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/testwasmlib/go/testwasmlib
github.com/iotaledger/wasp/contracts/wasm/tokenregistry/go/tokenregistryimpl imports
	github.com/iotaledger/wasp/contracts/wasm/tokenregistry/go/tokenregistry: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/tokenregistry/go/tokenregistry
github.com/iotaledger/wasp/contracts/wasm/corecontracts/test tested by
	github.com/iotaledger/wasp/contracts/wasm/corecontracts/test.test imports
	github.com/iotaledger/wasp/contracts/wasm/corecontracts/go/corecontracts: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/corecontracts/go/corecontracts
github.com/iotaledger/wasp/contracts/wasm/erc20/test tested by
	github.com/iotaledger/wasp/contracts/wasm/erc20/test.test imports
	github.com/iotaledger/wasp/contracts/wasm/erc20/go/erc20: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/erc20/go/erc20
github.com/iotaledger/wasp/contracts/wasm/gascalibration/executiontime/test tested by
	github.com/iotaledger/wasp/contracts/wasm/gascalibration/executiontime/test.test imports
	github.com/iotaledger/wasp/contracts/wasm/gascalibration/executiontime/go/executiontime: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/gascalibration/executiontime/go/executiontime
github.com/iotaledger/wasp/contracts/wasm/gascalibration/memory/test tested by
	github.com/iotaledger/wasp/contracts/wasm/gascalibration/memory/test.test imports
	github.com/iotaledger/wasp/contracts/wasm/gascalibration/memory/go/memory: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/gascalibration/memory/go/memory
github.com/iotaledger/wasp/contracts/wasm/gascalibration/storage/test tested by
	github.com/iotaledger/wasp/contracts/wasm/gascalibration/storage/test.test imports
	github.com/iotaledger/wasp/contracts/wasm/gascalibration/storage/go/storage: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/gascalibration/storage/go/storage
github.com/iotaledger/wasp/contracts/wasm/helloworld/test tested by
	github.com/iotaledger/wasp/contracts/wasm/helloworld/test.test imports
	github.com/iotaledger/wasp/contracts/wasm/helloworld/go/helloworld: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/helloworld/go/helloworld
github.com/iotaledger/wasp/contracts/wasm/schemacomment/test tested by
	github.com/iotaledger/wasp/contracts/wasm/schemacomment/test.test imports
	github.com/iotaledger/wasp/contracts/wasm/schemacomment/go/schemacomment: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/schemacomment/go/schemacomment
github.com/iotaledger/wasp/contracts/wasm/timestamp/test tested by
	github.com/iotaledger/wasp/contracts/wasm/timestamp/test.test imports
	github.com/iotaledger/wasp/contracts/wasm/timestamp/go/timestamp: cannot find module providing package github.com/iotaledger/wasp/contracts/wasm/timestamp/go/timestamp

</details>

**Comment by rarkins** [comment, marked as answer]
This is a general Renovate topic, not related directly to the app I think. Next step would be to test if the same type of error occurs when run against the CLI.

**Reply by rarkins** [reply]
If I clone your repo, check out commit `8cd7671fb6b06e40b726114302ea5e9699429e5d`, then run `go get -d -t ./...`, then I get the same error. This is on a Mac, non-Dockerized, so quite a different environment to what Renovate runs in the hosted app.

You'll need to work out what exact command(s) _do_ work on a clean machine, and then we can tell you whether that's possible to support in Renovate or not. Perhaps some of those modules need compiling first, for example?

**Reply by jorgemmsilva** [reply]
Thank you for the fast response. You're right.
Those "no go.mod" directories have to be compiled before `go get ./...` works :disappointed: .

Is there a way to configure renovate so that it only runs "go get" on target directories? instead of `./...`

**Reply by rarkins** [reply]
Right now there isn't a way, but it certainly is the type of thing which could be made configurable. Would it be something like `go get .`?

**Reply by jorgemmsilva** [reply]
for our case, ideally it would be:
```
go get -d -t . ./tools/...
```

**Reply by jorgemmsilva** [reply]
Any updates on this? we really need to fix this or disable renovate, as it keeps creating broken PRs :confused: .

I can try contributing, it should be a fairly straightforward change, but I'm not familiar with the renovate repo at all.

The configuration could look something like this:
![Selection_002](https://user-images.githubusercontent.com/1187222/216048978-b0173cd1-f1df-43f6-b0c7-7a288da9c830.png)
But I don't see this pattern used for any other language.
Also, how to obtain that value during `updateArtifacts`? :thinking:

**Reply by rarkins** [reply]
You can name it `goGetDirs` so it's a bit shorter. You'd pass it through similar to this one for composer: https://github.com/renovatebot/renovate/blob/fabb1bba0828c11c81513958ef3a9226880d6c76/lib/modules/manager/types.ts#L34

Also be sure that it's quoted using `shlex` (see https://github.com/renovatebot/renovate/blob/fabb1bba0828c11c81513958ef3a9226880d6c76/lib/modules/manager/composer/artifacts.ts#L2 for an example) so that it doesn't open up command injection attacks.

**Reply by rarkins** [reply]
Don't set it within `golang`, instead it should be defined within lib/config/options and a default there. Only the `gomod` manager would use it.
